File: packages/go-server/internal/handlers/eval_relay.go
package handlers

import (
	"bytes"
	"encoding/json"
	"io"
	"net/http"
	"os"
	"sync"
	"time"
)

// ── Server-side eval relay ──────────────────────────────────────────────────
//
// This server does NOT evaluate input — it is a thin relay between the panel
// and the compiled Go engine (packages/eval-engine, a separate binary the
// captain runs at PARLAY_EVAL_ADDR, default 127.0.0.1:4343). This is the Go
// port of packages/server/src/eval-relay.ts; like the TS original it contains
// ZERO command-matching logic, only transport. Evaluation running as compiled
// RE2 in Go — not interpreted JS in this server — is the whole point.
//
//   up:   panel POST /api/chat/eval        ──▶ engine POST /eval
//   down: engine response .actions          ──▶ broadcast input_action → device
//   fire: engine-owned submit timer          ──▶ POST /api/chat/eval-push
//
// Input evaluation is pure server-side and unconditional — every keystroke is
// relayed to the engine; there is no local fallback and no enable flag.

// evalEngineURL is where the compiled engine listens. Overridable for tests
// and non-standard deployments; the production default matches the engine's
// own coded default (packages/eval-engine/main.go's PARLAY_EVAL_ADDR).
func evalEngineURL() string {
	if v := os.Getenv("PARLAY_EVAL_ENGINE_URL"); v != "" {
		return v
	}
	return "http://127.0.0.1:4343"
}

// evalRelay holds the per-process state the two eval routes share: the
// streamId → deviceId routing table, so a later server-owned submit fire
// (which arrives on /eval-push carrying only a streamId) can be routed back
// to the owning panel. In-memory is fine: an engine restart drops armed
// timers anyway, and the panel re-arms on the next eval. Mirrors the TS
// relay's streamDevice map.
type evalRelay struct {
	streams sync.Map // streamId -> deviceId
}

func newEvalRelay() *evalRelay { return &evalRelay{} }

func (r *evalRelay) setStream(streamID, device string) { r.streams.Store(streamID, device) }

func (r *evalRelay) deviceFor(streamID string) (string, bool) {
	v, ok := r.streams.Load(streamID)
	if !ok {
		return "", false
	}
	return v.(string), true
}

// evalUpRequest is what this server forwards to the engine's POST /eval — the
// same body the panel sent, normalized to the engine's documented fields
// (packages/eval-engine/engine.go's EvalRequest). `device` is stripped before
// forwarding: it is panel-side routing, not something the engine knows about.
type evalUpRequest struct {
	StreamID     string `json:"streamId"`
	Version      int64  `json:"version"`
	Text         string `json:"text"`
	Cursor       any    `json:"cursor"`
	Reason       string `json:"reason"`
	VoiceEnabled bool   `json:"voiceEnabled"`
	Tabs         []any  `json:"tabs"`
}

// evalEnvelope is the engine's EvalResponse, forwarded to the wire with the
// timing the relay measured added. The verbs inside `actions` are opaque here.
type evalEnvelope struct {
	V            int64  `json:"v"`
	StreamID     string `json:"streamId"`
	Seq          int64  `json:"seq"`
	BaseVersion  int64  `json:"baseVersion"`
	Actions      []any  `json:"actions"`
	EngineEvalNs int64  `json:"engineEvalNs"`
	Fired        string `json:"fired"`
}

// relayTiming mirrors the TS relay's RelayTiming, surfaced to the panel so it
// can show compiled-eval time vs. total round-trip.
type relayTiming struct {
	EngineEvalNs int64 `json:"engineEvalNs"`
	RelayMs      int64 `json:"relayMs"`
	ServerOwned  bool  `json:"serverOwned,omitempty"`
}

// handleEval implements POST /api/chat/eval — the up-channel. It relays the
// panel's input snapshot to the compiled engine, then pushes the engine's
// actions to the owning device as one `input_action` event.
func handleEval(relay *evalRelay, hub *Hub) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			methodNotAllowed(w, http.MethodPost)
			return
		}
		var body struct {
			StreamID     string `json:"streamId"`
			Version      int64  `json:"version"`
			Text         string `json:"text"`
			Cursor       any    `json:"cursor"`
			Reason       string `json:"reason"`
			VoiceEnabled bool   `json:"voiceEnabled"`
			Tabs         []any  `json:"tabs"`
			Device       string `json:"device"`
		}
		if !decodeJSON(w, r, &body) {
			return
		}
		device := body.Device
		if device == "" {
			writeStatusError(w, http.StatusBadRequest, "device required")
			return
		}
		streamID := body.StreamID
		if streamID == "" {
			streamID = "eval-" + device + "-main"
		}
		relay.setStream(streamID, device)

		up := evalUpRequest{
			StreamID:     streamID,
			Version:      body.Version,
			Text:         body.Text,
			Cursor:       body.Cursor,
			Reason:       body.Reason,
			VoiceEnabled: body.VoiceEnabled,
			Tabs:         body.Tabs,
		}
		payload, _ := json.Marshal(up)

		t0 := time.Now()
		req, err := http.NewRequestWithContext(r.Context(), http.MethodPost, evalEngineURL()+"/eval", bytes.NewReader(payload))
		if err != nil {
			writeStatusError(w, http.StatusInternalServerError, err.Error())
			return
		}
		req.Header.Set("Content-Type", "application/json")
		client := &http.Client{Timeout: 2 * time.Second}
		resp, err := client.Do(req)
		if err != nil {
			writeStatusError(w, http.StatusBadGateway, "engine unreachable: "+err.Error())
			return
		}
		defer resp.Body.Close()
		if resp.StatusCode != http.StatusOK {
			writeStatusError(w, http.StatusBadGateway, "engine "+resp.Status)
			return
		}
		raw, err := io.ReadAll(resp.Body)
		if err != nil {
			writeStatusError(w, http.StatusBadGateway, "engine read: "+err.Error())
			return
		}
		var env evalEnvelope
		if err := json.Unmarshal(raw, &env); err != nil {
			writeStatusError(w, http.StatusBadGateway, "engine bad response: "+err.Error())
			return
		}
		relayMs := time.Since(t0).Milliseconds()

		timing := relayTiming{EngineEvalNs: env.EngineEvalNs, RelayMs: relayMs}
		hub.broadcastToDevice(device, eventInputAction, map[string]any{
			"v":           env.V,
			"streamId":    env.StreamID,
			"seq":         env.Seq,
			"baseVersion": env.BaseVersion,
			"actions":     env.Actions,
			"timing":      timing,
		})

		// Return the envelope + timing synchronously too, so a curl/Interceptor
		// probe and the client's latency overlay can read it without the SSE hop.
		writeJSON(w, map[string]any{
			"ok":          true,
			"v":           env.V,
			"streamId":    env.StreamID,
			"seq":         env.Seq,
			"baseVersion": env.BaseVersion,
			"actions":     env.Actions,
			"timing":      timing,
		})
	}
}

// handleEvalPush implements POST /api/chat/eval-push — the down-channel for
// SERVER-OWNED submit fires. The engine calls this when its per-stream timer
// elapses; we look up the owning device and push the submitNow over SSE.
func handleEvalPush(relay *evalRelay, hub *Hub) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			methodNotAllowed(w, http.MethodPost)
			return
		}
		var body struct {
			StreamID    string `json:"streamId"`
			Seq         int64  `json:"seq"`
			BaseVersion int64  `json:"baseVersion"`
			V           int64  `json:"v"`
			Action      any    `json:"action"`
		}
		if !decodeJSON(w, r, &body) {
			return
		}
		device, ok := relay.deviceFor(body.StreamID)
		if !ok {
			writeStatusError(w, http.StatusNotFound, "unknown stream")
			return
		}
		actions := []any{}
		if body.Action != nil {
			actions = append(actions, body.Action)
		}
		hub.broadcastToDevice(device, eventInputAction, map[string]any{
			"v":           orDefault(body.V, 1),
			"streamId":    body.StreamID,
			"seq":         body.Seq,
			"baseVersion": body.BaseVersion,
			"actions":     actions,
			"timing":      relayTiming{ServerOwned: true},
		})
		writeJSON(w, map[string]any{"ok": true})
	}
}

func orDefault(v, def int64) int64 {
	if v == 0 {
		return def
	}
	return v
}

File: packages/server/src/eval-relay.ts
import { broadcastToDevice, CORS } from "./sse"

// ── Server-side eval relay (feat/server-side-eval) ─────────────────────────────
//
// PURE server-side input evaluation. The TS server does NOT evaluate — it is a
// thin relay between the client and the COMPILED Go engine (tools/cli/internal/evalengine).
// This is deliberate: the captain's whole point is that evaluation runs as
// compiled RE2 in Go, not interpreted JS. Doing the matching here in bun would
// defeat the purpose. So this file contains ZERO command-matching logic — only
// transport.
//
//   up:   client POST /api/chat/eval  ──▶  this relay  ──▶  Go /eval (HTTP)
//   down: Go /eval response.actions    ──▶  broadcastToDevice(input_action) ─▶ SSE
//   fire: Go server-owned submit timer ──▶  POST /api/chat/eval-push ─▶ SSE
//
// Input evaluation is PURE server-side and unconditional — there is no local
// pipeline and no enable flag. Every keystroke is relayed to the Go engine.

const EVAL_ENGINE_URL =
  process.env.PARLAY_EVAL_ENGINE_URL ?? "http://127.0.0.1:4343"

// The action-protocol envelope the client dispatcher expects. Built from the Go
// engine's response; the relay is agnostic to the verbs inside.
interface EvalEnvelope {
  v: number
  streamId: string
  seq: number
  baseVersion: number
  actions: unknown[]
  engineEvalNs: number
  fired: string
}

// Relay round-trip timing surfaced to the client so it can display the
// compiled-eval-time vs. total-round-trip comparison the captain wants to see.
interface RelayTiming {
  engineEvalNs: number   // pure compiled eval time (from Go)
  relayMs: number        // TS→Go→TS HTTP round-trip as seen by the relay
}

export async function handleEvalRequest(
  req: Request,
  pathname: string,
): Promise<Response | null> {
  // POST /api/chat/eval — the up-channel. Delegates to the compiled Go engine
  // and relays its actions over the device-scoped SSE.
  if (req.method === "POST" && pathname === "/api/chat/eval") {
    let body: {
      streamId?: string
      version?: number
      text?: string
      cursor?: { anchor: number; active: number }
      reason?: string
      voiceEnabled?: boolean
      tabs?: { id: string; name: string }[]
      device?: string
    }
    try {
      body = await req.json()
    } catch {
      return json({ error: "bad request" }, 400)
    }
    const device = String(body.device ?? "").trim()
    if (!device) return json({ error: "device required" }, 400)
    const streamId = String(body.streamId ?? `eval-${device}-main`)

    const engineReq = {
      streamId,
      version: Number(body.version ?? 0),
      text: String(body.text ?? ""),
      cursor: body.cursor ?? { anchor: 0, active: 0 },
      reason: String(body.reason ?? "input"),
      voiceEnabled: body.voiceEnabled === true,
      tabs: Array.isArray(body.tabs) ? body.tabs : [],
    }

    // Remember which device owns this stream so a later server-owned submit fire
    // (which arrives on /eval-push with only a streamId) can be routed back.
    streamDevice.set(streamId, device)

    const t0 = performance.now()
    let env: EvalEnvelope
    try {
      const r = await fetch(`${EVAL_ENGINE_URL}/eval`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(engineReq),
        signal: AbortSignal.timeout(2_000),
      })
      if (!r.ok) return json({ error: `engine ${r.status}` }, 502)
      env = (await r.json()) as EvalEnvelope
    } catch (e) {
      // Engine unreachable — fail loud but non-fatal. No local fallback exists;
      // commands simply don't fire for this keystroke until the engine returns.
      return json({ error: "engine unreachable", detail: String(e) }, 502)
    }
    const relayMs = performance.now() - t0

    const timing: RelayTiming = { engineEvalNs: env.engineEvalNs, relayMs }

    // Relay the actions over the existing device-scoped SSE as one input_action
    // event carrying the full envelope. The client dispatcher validates seq /
    // baseVersion / ttl before applying (dispatcher.ts).
    const matched = broadcastToDevice(device, "input_action", {
      v: env.v,
      streamId: env.streamId,
      seq: env.seq,
      baseVersion: env.baseVersion,
      actions: env.actions,
      timing,
    })

    // Also return the envelope+timing synchronously so a curl/Interceptor probe
    // (and the client's latency overlay) can read it without the SSE hop.
    return json({ ok: true, sseClients: matched, ...env, timing }, 200)
  }

  // POST /api/chat/eval-push — the down-channel for SERVER-OWNED submit fires.
  // The Go engine calls this when its per-stream 1s timer elapses; we look up the
  // owning device and push the submitNow over SSE. This is the leg that carries
  // the network-race cost: the fire is already ~1 round-trip stale by the time it
  // reaches the client, which re-verifies the tail before actually sending.
  if (req.method === "POST" && pathname === "/api/chat/eval-push") {
    let body: { streamId?: string; seq?: number; baseVersion?: number; v?: number; action?: unknown }
    try {
      body = await req.json()
    } catch {
      return json({ error: "bad request" }, 400)
    }
    const streamId = String(body.streamId ?? "")
    const device = streamDevice.get(streamId)
    if (!device) return json({ error: "unknown stream", streamId }, 404)

    const matched = broadcastToDevice(device, "input_action", {
      v: body.v ?? 1,
      streamId,
      seq: body.seq ?? 0,
      baseVersion: body.baseVersion ?? 0,
      actions: body.action ? [body.action] : [],
      timing: { serverOwnedFire: true },
    })
    return json({ ok: true, sseClients: matched }, 200)
  }

  return null
}

// streamId → deviceId, so a server-owned submit fire (which only knows the
// stream) can be routed to the right SSE device. In-memory is fine: a Pulse
// restart drops armed timers anyway, and the client re-arms on the next eval.
const streamDevice = new Map<string, string>()

function json(data: unknown, status: number): Response {
  return new Response(JSON.stringify(data), {
    status,
    headers: { "Content-Type": "application/json", ...CORS },
  })
}
